File: csharp/Slimmed/PublishingHouse.Api/Controllers/BooksController.cs
using Microsoft.AspNetCore.Mvc;
using PublishingHouse.Api.Requests;
using PublishingHouse.Application.Books;
using PublishingHouse.Books;
using PublishingHouse.Books.Authors;
using PublishingHouse.Books.Draft;
using PublishingHouse.Books.Entities;
using PublishingHouse.Core.Validation;
using PublishingHouse.Core.ValueObjects;

namespace PublishingHouse.Api.Controllers;

using static DraftCommand;

[Route("api/[controller]")]
public class BooksController: Controller
{
    private readonly IBooksService booksService;
    private readonly IBookQueryService booksQueryService;

    public BooksController(IBooksService booksService, IBookQueryService booksQueryService)
    {
        this.booksService = booksService;
        this.booksQueryService = booksQueryService;
    }

    [HttpPost]
    public async Task<IActionResult> CreateDraft([FromBody] CreateDraftRequest request, CancellationToken ct)
    {
        var bookId = Guid.NewGuid();

        var (title, author, publisherId, edition, genre) = request;

        author.AssertNotNull();

        await booksService.CreateDraft(
            new CreateDraft(
                new BookId(bookId),
                new Title(title.AssertNotEmpty()),
                new AuthorIdOrData(
                    author.AuthorId.HasValue ? new AuthorId(author.AuthorId.Value) : null,
                    author.FirstName != null ? new AuthorFirstName(author.FirstName) : null,
                    author.LastName != null ? new AuthorLastName(author.LastName) : null
                ),
                new PublisherId(publisherId.AssertNotEmpty()),
                new PositiveInt(edition.GetValueOrDefault()),
                genre != null ? new Genre(genre) : null
            ),
            ct
        );

        return Created($"/api/books/{bookId}", bookId);
    }

    [HttpPost("{id}/chapters")]
    public async Task<IActionResult> AddChapter([FromRoute] Guid id, [FromBody] AddChapterRequest request, CancellationToken ct)
    {
        var (title, content) = request;

        await booksService.AddChapter(
            new AddChapter(
                new BookId(id),
                new ChapterTitle(title.AssertNotEmpty()),
                content != null ? new ChapterContent(content): ChapterContent.Empty
            ),
            ct
        );

        return NoContent();
    }

    [HttpPost("{id}/move-to-editing")]
    public async Task<IActionResult> MoveToEditing([FromRoute] Guid id, CancellationToken ct)
    {
        await booksService.MoveToEditing(
            new MoveToEditing(new BookId(id)),
            ct
        );

        return NoContent();
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> FindDetailsById([FromRoute] Guid id, CancellationToken ct) =>
        await booksQueryService.FindDetailsById(new BookId(id), ct) is { } result ? Ok(result) : NotFound();
}

File: java/src/main/java/io/eventdriven/slimdownaggregates/original/api/controllers/BooksController.java
package io.eventdriven.slimdownaggregates.original.api.controllers;

import io.eventdriven.slimdownaggregates.original.api.requests.AddChapterRequest;
import io.eventdriven.slimdownaggregates.original.api.requests.CreateDraftRequest;
import io.eventdriven.slimdownaggregates.original.application.books.BooksQueryService;
import io.eventdriven.slimdownaggregates.original.application.books.BooksService;
import io.eventdriven.slimdownaggregates.original.application.books.commands.AddChapterCommand;
import io.eventdriven.slimdownaggregates.original.application.books.commands.CreateDraftCommand;
import io.eventdriven.slimdownaggregates.original.application.books.commands.MoveToEditingCommand;
import io.eventdriven.slimdownaggregates.original.domain.books.authors.AuthorIdOrData;
import io.eventdriven.slimdownaggregates.original.domain.books.entities.*;
import io.eventdriven.slimdownaggregates.original.infrastructure.valueobjects.PositiveInt;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.UUID;

@RestController
@RequestMapping("/api/books")
public class BooksController {

  private final BooksService booksService;
  private final BooksQueryService booksQueryService;

  public BooksController(BooksService booksService, BooksQueryService booksQueryService) {
    this.booksService = booksService;
    this.booksQueryService = booksQueryService;
  }

  @PostMapping
  public ResponseEntity<UUID> createDraft(@RequestBody CreateDraftRequest request) throws URISyntaxException {
    var bookId = UUID.randomUUID();

    var author = request.author();

    booksService.createDraft(
      new CreateDraftCommand(
        new BookId(bookId),
        new Title(request.title()),
        new AuthorIdOrData(
          author.authorId() != null ? new AuthorId(author.authorId()) : null,
          author.firstName() != null ? new AuthorFirstName(author.firstName()) : null,
          author.lastName() != null ? new AuthorLastName(author.lastName()) : null
        ),
        new PublisherId(request.publisherId()),
        new PositiveInt(request.edition()),
        request.genre() != null ? new Genre(request.genre()) : null
      )
    );

    return ResponseEntity
      .created(new URI("api/books/%s".formatted(bookId)))
      .build();
  }

  @PostMapping("/{id}/chapters")
  public ResponseEntity<Void> addChapter(@PathVariable UUID id, @RequestBody AddChapterRequest request) {
    booksService.addChapter(
      new AddChapterCommand(
        new BookId(id),
        new ChapterTitle(request.title()),
        request.content() != null ? new ChapterContent(request.content()) : ChapterContent.empty
      )
    );

    return ResponseEntity.noContent().build();
  }

  @PostMapping("/{id}/move-to-editing")
  public ResponseEntity<Void> moveToEditing(@PathVariable UUID id) {
    booksService.moveToEditing(
      new MoveToEditingCommand(new BookId(id))
    );

    return ResponseEntity.noContent().build();
  }

  @GetMapping("/{id}")
  public ResponseEntity<?> findDetailsById(@PathVariable UUID id) {
    var result = booksQueryService.findDetailsById(new BookId(id));

    return result.map(ResponseEntity::ok)
      .orElse(ResponseEntity.notFound().build());
  }
}

